service/dynamodb/dynamodbattribute/encode.go

```diff
@@ -0,0 +1,465 @@
+		*av.BOOL = typed
+	case string:
+		if err := e.encodeString(av, v); err != nil {
+			return err
+		}
+	default:
+		// Fallback to encoding numbers, will return invalid type if not supported
+		if err := e.encodeNumber(av, v); err != nil {
+			return err
+		}
+		if fieldTag.AsString && av.NULL == nil && av.N != nil {
+			av.S = av.N
+			av.N = nil
+		}
+	}
+
+	return nil
+}
+
+func (e *Encoder) encodeNumber(av *dynamodb.AttributeValue, v reflect.Value) error {
+	if used, err := tryMarshaler(av, v); used {
+		return err
+	}
+
+	var out string
+	switch typed := v.Interface().(type) {
+	case int:
+		out = encodeInt(int64(typed))
+	case int8:
+		out = encodeInt(int64(typed))
+	case int16:
+		out = encodeInt(int64(typed))
+	case int32:
+		out = encodeInt(int64(typed))
+	case int64:
+		out = encodeInt(typed)
+	case uint:
+		out = encodeUint(uint64(typed))
+	case uint8:
+		out = encodeUint(uint64(typed))
+	case uint16:
+		out = encodeUint(uint64(typed))
+	case uint32:
+		out = encodeUint(uint64(typed))
+	case uint64:
+		out = encodeUint(typed)
+	case float32:
+		out = encodeFloat(float64(typed))
+	case float64:
+		out = encodeFloat(typed)
+	default:
+		return &unsupportedMarshalTypeError{Type: v.Type()}
+	}
+
+	av.N = &out
+
+	return nil
+}
+
+func (e *Encoder) encodeString(av *dynamodb.AttributeValue, v reflect.Value) error {
+	if used, err := tryMarshaler(av, v); used {
+		return err
+	}
+
+	switch typed := v.Interface().(type) {
+	case string:
+		if len(typed) == 0 && e.NullEmptyString {
+			encodeNull(av)
+		} else {
+			av.S = &typed
+		}
+	default:
+		return &unsupportedMarshalTypeError{Type: v.Type()}
+	}
+
+	return nil
+}
+
+func encodeInt(i int64) string {
+	return strconv.FormatInt(i, 10)
+}
+func encodeUint(u uint64) string {
+	return strconv.FormatUint(u, 10)
+}
+func encodeFloat(f float64) string {
+	return strconv.FormatFloat(f, 'f', -1, 64)
+}
+func encodeNull(av *dynamodb.AttributeValue) {
+	t := true
+	*av = dynamodb.AttributeValue{NULL: &t}
+}
+
+func valueElem(v reflect.Value) reflect.Value {
+	switch v.Kind() {
+	case reflect.Interface, reflect.Ptr:
+		for v.Kind() == reflect.Interface || v.Kind() == reflect.Ptr {
+			v = v.Elem()
+		}
+	}
+
+	return v
+}
+
+func emptyValue(v reflect.Value) bool {
+	switch v.Kind() {
+	case reflect.Array, reflect.Map, reflect.Slice, reflect.String:
+		return v.Len() == 0
+	case reflect.Bool:
+		return !v.Bool()
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		return v.Int() == 0
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uintptr:
+		return v.Uint() == 0
+	case reflect.Float32, reflect.Float64:
+		return v.Float() == 0
+	case reflect.Interface, reflect.Ptr:
+		return v.IsNil()
+	}
+	return false
+}
+
+func tryMarshaler(av *dynamodb.AttributeValue, v reflect.Value) (bool, error) {
+	if v.Kind() != reflect.Ptr && v.Type().Name() != "" && v.CanAddr() {
+		v = v.Addr()
+	}
+
+	if v.Type().NumMethod() == 0 {
```

Comment:
This is an optimization to only look if the type implements the Marshaler interface if there are methods on the type. The check for num methods allows us to skip the type assertion if we know it will never match.